ci-operator/step-registry/ipi/install/install/in-bastion/ipi-install-install-in-bastion-commands.sh

```diff
@@ -213,6 +213,11 @@ set -o errexit
 # debug
 echo "the content of ${dir} in local container "
 ls -ltra ${dir}
+# ensure installation does succeed
+if ! grep -q "level=info msg=\"Install complete!\"" "${dir}"/.openshift_install.log; then
```

Comment:
@jinyunma I added such check because I happened noticing "Installer exit with code 0" although the installation failed, see [here](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/46111/rehearse-46111-periodic-ci-openshift-verification-tests-master-installer-rehearse-4.15-installer-rehearse-debug/1760125881650515968). Please advise, thanks!

Reply:
Yes, the first run yesterday failed, but the subsequent run got successful installation. Anyway, I tried similar `COMMAND 2>&1 | grep...` and it seems even COMMAND exit with 1, grep would exit with 0. 

```
[jiwei@jiwei ~]$ date abc 2>&1 | grep --line-buffered -v 'password\|X-Auth-Token\|UserData:'
date: invalid date ‘abc’
[jiwei@jiwei ~]$ echo $?
0
[jiwei@jiwei ~]$ date abc 2>&1
date: invalid date ‘abc’
[jiwei@jiwei ~]$ echo $?
1
[jiwei@jiwei ~]$

Reply:
My local testing sounds working well.
```
[jianlinliu@jialiu-laptop ~]$ cat test.sh 
ssh jialiu-rhel8 "ls /tmp | grep --line-buffered -v 'password\|X-Auth-Token\|UserData:'" &
wait "$!"
ret="$?"
echo "Installer exit with code $ret"
[jianlinliu@jialiu-laptop ~]$ sh test.sh 
Warning: Permanently added '10.0.76.82' (ED25519) to the list of known hosts.
systemd-private-68efcaecbf024cbd8893caf8b1096b91-chronyd.service-PhW9BT
Installer exit with code 0

[jianlinliu@jialiu-laptop ~]$ cat test.sh 
ssh jialiu-rhel8 "ls /no-existing | grep --line-buffered -v 'password\|X-Auth-Token\|UserData:'" &
wait "$!"
ret="$?"
echo "Installer exit with code $ret"
[jianlinliu@jialiu-laptop ~]$ sh test.sh 
Warning: Permanently added '10.0.76.82' (ED25519) to the list of known hosts.
ls: cannot access '/no-existing': No such file or directory
Installer exit with code 1
```

Reply:
@jianlinliu Please try with "ls /no-existing **2>&1** | grep --line-buffered -v 'password\|X-Auth-Token\|UserData:'"

Reply:
indeed appending grep command seems always exiting with 0.
```
[jima@mylaptop ~]$ ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o ServerAliveInterval=60 -o ServerAliveCountMax=2 fedora@10.0.79.191  "ls /no-existing 2>&1 | grep --line-buffered -v 'password\|X-Auth-Token\|UserData:'" &
[1] 7193
[jima@mylaptop ~]$ wait "$!"; ret="$?"; echo $ret
Warning: Permanently added '10.0.79.191' (ED25519) to the list of known hosts.
ls: cannot access '/no-existing': No such file or directory
[1]+  Done                    ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o ServerAliveInterval=60 -o ServerAliveCountMax=2 fedora@10.0.79.191 "ls /no-existing 2>&1 | grep --line-buffered -v 'password\|X-Auth-Token\|UserData:'"
0

Remove grep command:
[jima@mylaptop ~]$ ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o ServerAliveInterval=60 -o ServerAliveCountMax=2 fedora@10.0.79.191  "ls /no-existing 2>&1" &
[1] 7164
[jima@mylaptop ~]$ wait "$!"; ret="$?"; echo $ret
Warning: Permanently added '10.0.79.191' (ED25519) to the list of known hosts.
ls: cannot access '/no-existing': No such file or directory
[1]+  Exit 2                  ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o ServerAliveInterval=60 -o ServerAliveCountMax=2 fedora@10.0.79.191 "ls /no-existing 2>&1"
2
```
But from .opensihft-install.log, seems that installation is still ongoing, but not failed, not sure why ssh command exited.

Reply:
Ah, I got the root cause, we can fix it by changing the following line:
```
run_ssh_cmd "${SSH_PRIV_KEY_PATH}" "${BASTION_SSH_USER}" "${BASTION_IP}" "source ${REMOTE_ENV_FILE}; ${REMOTE_DIR}/openshift-install --dir='${REMOTE_INSTALL_DIR}' create cluster --log-level debug 2>&1 | grep --line-buffered -v 'password\|X-Auth-Token\|UserData:'" &
```
to 
```
run_ssh_cmd "${SSH_PRIV_KEY_PATH}" "${BASTION_SSH_USER}" "${BASTION_IP}" "source ${REMOTE_ENV_FILE}; ${REMOTE_DIR}/openshift-install --dir='${REMOTE_INSTALL_DIR}' create cluster --log-level debug 2>&1" | grep --line-buffered -v 'password\|X-Auth-Token\|UserData:' &
```

Reply:
Two CI jobs both failed at installation stage, 
[periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-oidc-arm-f14](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/logs/periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-oidc-arm-f14/1760340976368357376), failed at ipi-install-install
```
level=error msg=Error checking cluster operator Progressing status: "context deadline exceeded"
level=error msg=These cluster operators were not stable: [image-registry]
Installer exit with code 7
```
[periodic-ci-openshift-verification-tests-master-installation-nightly-4.15-azure-ipi-oidc-managed-identity-user-defined-f14](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/logs/periodic-ci-openshift-verification-tests-master-installation-nightly-4.15-azure-ipi-oidc-managed-identity-user-defined-f14/1760369144647127040), step ipi-install-install-in-bastion is passed, installation failed actually.
```
level=debug msg=These cluster operators were stable: [authentication, baremetal, cloud-controller-manager, cloud-credential, cluster-autoscaler, config-operator, console, control-plane-machine-set, csi-snapshot-controller, dns, etcd, ingress, insights, kube-apiserver, kube-controller-manager, kube-scheduler, kube-storage-version-migrator, machine-api, machine-approver, machine-config, marketplace, monitoring, network, node-tuning, openshift-apiserver, openshift-controller-manager, openshift-samples, operator-lifecycle-manager, operator-lifecycle-manager-catalog, operator-lifecycle-manager-packageserver, service-ca, storage]
level=error msg=These cluster operators were not stable: [image-registry]
Installer exit with code 0
```
step `ipi-install-install` exit with expected code, because it `set -o pipefail` .

maybe we also need to add `set -o pipefail` in ssh command to install cluster in step `ipi-install-install-in-bastion`?
```
[jima@mylaptop ~]$ ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o ServerAliveInterval=60 -o ServerAliveCountMax=2 fedora@10.0.79.191  "set -o pipefail; ls /no-existing 2>&1 | grep --line-buffered -v 'password\|X-Auth-Token\|UserData:'" &
[1] 8185
[jima@mylaptop ~]$ wait "$!"; ret="$?"; echo $ret
Warning: Permanently added '10.0.79.191' (ED25519) to the list of known hosts.
ls: cannot access '/no-existing': No such file or directory
[1]+  Exit 2                  ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o ServerAliveInterval=60 -o ServerAliveCountMax=2 fedora@10.0.79.191 "set -o pipefail; ls /no-existing 2>&1 | grep --line-buffered -v 'password\|X-Auth-Token\|UserData:'"
2
```

Reply:
@jinyunma I updated with Jianlin's suggestion. WDYT?

Reply:
> step ipi-install-install exit with expected code, because it set -o pipefail .
> maybe we also need to add set -o pipefail in ssh command to install cluster in step ipi-install-install-in-bastion

Yeah, this is another solution. Two solutions:
1. run pipe command (`grep`) locally instead of in remote ssh command.
2. run pipe command (`grep`) in remote ssh command with `pipefail` enabled.

Reply:
Both are okay for me.